<a href="https://colab.research.google.com/github/Ruhan-Saad-Dave/My-Projects/blob/main0/Student_Performance_prediction_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

# understanding the problem

This dataset revolves around the GPA of a student while tracking their demographics, study habits, parental involvement, extracurricular activities, and academic performance.

The goal is to find a unique relations between the different aspects and the GPA, hence will be used for prediction.

This project will be usefull in predicting the expected GPA of a student based on their demographics, study habits, parental involvement, extracurricular activities, and academic performance.

Source: https://www.kaggle.com/datasets/rabieelkharoua/students-performance-dataset

# Features of dataset

* StudentID: A unique identifier assigned to each student (1001 to 3392).
* Age: The age of the students ranges from 15 to 18 years.
* Gender: Gender of the students, where 0 represents Male and 1 represents Female.
* Ethnicity: The ethnicity of the students, coded as follows:
  * 0: Caucasian
  * 1: African American
  * 2: Asian
  * 3: Other
* ParentalEducation: The education level of the parents, coded as follows:
  * 0: None
  * 1: High School
  * 2: Some College
  * 3: Bachelor's
  * 4: Higher
* StudyTimeWeekly: Weekly study time in hours, ranging from 0 to 20.
* Absences: Number of absences during the school year, ranging from 0 to 30.
* Tutoring: Tutoring status, where 0 indicates No and 1 indicates Yes.
* ParentalSupport: The level of parental support, coded as follows:
  * 0: None
  * 1: Low
  * 2: Moderate
  * 3: High
  * 4: Very High
* Extracurricular: Participation in extracurricular activities, where 0 indicates No and 1 indicates Yes.
* Sports: Participation in sports, where 0 indicates No and 1 indicates Yes.
* Music: Participation in music activities, where 0 indicates No and 1 indicates Yes.
* Volunteering: Participation in volunteering, where 0 indicates No and 1 indicates Yes.
* GPA: Grade Point Average on a scale from 2.0 to 4.0, influenced by study habits, parental involvement, and extracurricular activities.
* GradeClass: Classification of students' grades based on GPA:
  * 0: 'A' (GPA >= 3.5)
  * 1: 'B' (3.0 <= GPA < 3.5)
  * 2: 'C' (2.5 <= GPA < 3.0)
  * 3: 'D' (2.0 <= GPA < 2.5)
  * 4: 'F' (GPA < 2.0)

# Data preprocessing

Before using the dataset for predictions, we should check if there are any rows of information with empty values or wrong format.

In [184]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("Student_performance_data _.csv")
df.head()

StudentID  Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  \
0       1001   17       1          0                  2        19.833723   
1       1002   18       0          0                  1        15.408756   
2       1003   15       0          2                  3         4.210570   
3       1004   17       1          0                  3        10.028829   
4       1005   17       1          0                  2         4.672495   

   Absences  Tutoring  ParentalSupport  Extracurricular  Sports  Music  \
0         7         1                2                0       0      1   
1         0         0                1                0       0      0   
2        26         0                2                0       0      0   
3        14         0                3                1       0      0   
4        17         1                3                0       0      0   

   Volunteering       GPA  GradeClass  
0             0  2.929196         2.0  
1             0  3.042915         1.0  
2             0  0.112602         4.0  
3             0  2.054218         3.0  
4             0  1.288061         4.0

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   StudentID          2392 non-null   int64  
 1   Age                2392 non-null   int64  
 2   Gender             2392 non-null   int64  
 3   Ethnicity          2392 non-null   int64  
 4   ParentalEducation  2392 non-null   int64  
 5   StudyTimeWeekly    2392 non-null   float64
 6   Absences           2392 non-null   int64  
 7   Tutoring           2392 non-null   int64  
 8   ParentalSupport    2392 non-null   int64  
 9   Extracurricular    2392 non-null   int64  
 10  Sports             2392 non-null   int64  
 11  Music              2392 non-null   int64  
 12  Volunteering       2392 non-null   int64  
 13  GPA                2392 non-null   float64
 14  GradeClass         2392 non-null   float64
dtypes: float64(3), int64(12)
memory usage: 280.4 KB


In [ ]:
df.isnull().sum()

StudentID            0
Age                  0
Gender               0
Ethnicity            0
ParentalEducation    0
StudyTimeWeekly      0
Absences             0
Tutoring             0
ParentalSupport      0
Extracurricular      0
Sports               0
Music                0
Volunteering         0
GPA                  0
GradeClass           0
dtype: int64

From here we got to know that there is no row withblank values.

# Building models

In [43]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("Student_performance_data _.csv")
df.head()

StudentID  Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  \
0       1001   17       1          0                  2        19.833723   
1       1002   18       0          0                  1        15.408756   
2       1003   15       0          2                  3         4.210570   
3       1004   17       1          0                  3        10.028829   
4       1005   17       1          0                  2         4.672495   

   Absences  Tutoring  ParentalSupport  Extracurricular  Sports  Music  \
0         7         1                2                0       0      1   
1         0         0                1                0       0      0   
2        26         0                2                0       0      0   
3        14         0                3                1       0      0   
4        17         1                3                0       0      0   

   Volunteering       GPA  GradeClass  
0             0  2.929196         2.0  
1             0  3.042915         1.0  
2             0  0.112602         4.0  
3             0  2.054218         3.0  
4             0  1.288061         4.0

Here, we will use the columns(features): Age, Gender, Ethnicity, ParentalEducation, StudyTimeWeekly, Absences, Tutoring, ParentalSupport, Extracurricular, SPorts, Music, Volunteering.

And column (label) GPA, GradeClass for regression and classification respectively.

In [3]:
X = df.iloc[:, 1:-2].values
y1 = df.iloc[:, -2].values
y2 = df.iloc[:, -1].values

Splitting the dataset into training and testing set.

In [4]:
from sklearn.model_selection import train_test_split
X1_train, X1_test, y1_train, y1_test = train_test_split(X, y1, test_size = 0.2, random_state = 42)
X2_train, X2_test, y2_train, y2_test = train_test_split(X, y2, test_size = 0.2, random_state = 42)

## Regression

Here, we will build regression modles to predict the GPA stored in y1, and try to find the best possible model.

### Multiple linear regression

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
reg = LinearRegression()
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

0.9532471681022928

### Polynomial Regression

In [ ]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

poly_reg = PolynomialFeatures(degree = 2)
reg = LinearRegression()
X_poly = poly_reg.fit_transform(X1_train)
reg.fit(X_poly, y1_train)
y_pred = reg.predict(poly_reg.transform(X1_test))
r2_score(y1_test, y_pred)

0.9511391131209972

### Decision Tree Regression

In [ ]:
from sklearn.tree import DecisionTreeRegressor
reg = DecisionTreeRegressor(random_state = 42)
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

0.8658888351952292

### Random Forest Regression

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score
reg = RandomForestRegressor(n_estimators = 100, random_state = 42)
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

0.9268439109198162

### Model fine tuning

From all the regression algorithm we have used, Multiple linear regression had the best result. So we will try to fine tune the model for better results.

Parameters of LinearRegressor:
* fit_intercept : bool, default=True
Whether to calculate the intercept for this model. If set to False, no intercept will be used in calculations (i.e. data is expected to be centered).

* copy_X : bool, default=True
If True, X will be copied; else, it may be overwritten.

* n_jobs: int, default=None
The number of jobs to use for the computation. This will only provide speedup in case of sufficiently large problems, that is if firstly n_targets > 1 and secondly X is sparse or if positive is set to True. None means 1 unless in a joblib.parallel_backend context. -1 means using all processors. See Glossary for more details.

* positive : bool, default=False
When set to True, forces the coefficients to be positive. This option is only supported for dense arrays.

In [ ]:
#Original
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
reg = LinearRegression()
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

0.9532471681022928

In [ ]:
#fit_intercept = False
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
reg = LinearRegression(fit_intercept = False)
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

0.9180453068908785

In [ ]:
#copy_X = True
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
reg = LinearRegression(copy_X = True)
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

#Value of X will be changed if copy_X = False

0.9532471681022928

In [ ]:
#positive = True
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
reg = LinearRegression(positive = True)
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

0.08669024449420559

In [ ]:
#n_jobs
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
reg = LinearRegression(n_jobs = 1)
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

#value of n_jobs doesnt really effect the result

0.9532471681022928

After experimenting all the parameters, the default parameters gives the best result

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
reg = LinearRegression(fit_intercept = True, positive = False, copy_X = True, n_jobs = 1)
reg.fit(X1_train, y1_train)
y_pred = reg.predict(X1_test)
r2_score(y1_test, y_pred)

0.9532471681022928

### Conclusion

For regression prediction, we found that linear regression gives the best result with R2 Score of 0.9532

## Classification

Now we will do classification prediction for the GradeClass which is stored in y2.

### Logistic regression

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

sc_X = StandardScaler()
X_train = sc_X.fit_transform(X2_train)
X_test = sc_X.transform(X2_test)
reg = LogisticRegression(random_state = 42)
reg.fit(X_train, y2_train)
y_pred = reg.predict(X_test)
accuracy_score(y2_test, y_pred)

0.6826722338204593

### K Nearest Neighbors Classifiers

In [16]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
cls = KNeighborsClassifier(n_neighbors = 21)
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

0.6931106471816284

### Support Vector Classifier

In [37]:
from sklearn.svm import SVC
cls = SVC(kernel = "rbf", random_state = 42)
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

0.6889352818371608

### Naive Bayes

In [ ]:
from sklearn.naive_bayes import GaussianNB
cls = GaussianNB()
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

0.6325678496868476

### Decision Tree Classifier

In [ ]:
from sklearn.tree import DecisionTreeClassifier
cls = DecisionTreeClassifier(criterion = "entropy", random_state = 42)
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

0.6075156576200418

### Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
cls = RandomForestClassifier(n_estimators = 23, criterion = "entropy", random_state = 42)
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

0.6993736951983298

### Multi-layer Perceptron classifier. (Neural network)

In [176]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

cls = MLPClassifier(random_state = 42)
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


0.7473903966597077

### Model Fine Tuneing

From all the classification algorithms, MLP Classifier (Multi-layer Perceptron classifier.) performed the best, so now we will fine tune the model to give better results.

In [175]:
#original
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

cls = MLPClassifier(random_state = 42)
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


0.7473903966597077

In [ ]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

cls = MLPClassifier(random_state = 42, max_iter = 261)
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

### Conclusion

After testing out the different parameters, we have achieved the accuracy score of 0.7515

In [183]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

cls = MLPClassifier(random_state = 42, max_iter = 261, activation = "relu", solver = "adam",early_stopping = False)
cls.fit(X2_train, y2_train)
y_pred = cls.predict(X2_test)
accuracy_score(y2_test, y_pred)

/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (261) reached and the optimization hasn't converged yet.
  warnings.warn(


0.7515657620041754

# Final Conclusion

We are able to build a regression and classification models with 95.32% and 75.15% accuracy respectively. This shows that the features of the dataset have a strong relation with the GPA of the student.